### PR TITLE
Fixed #5711: serialize doesn't work with extra()

### PR DESCRIPTION
**Pull request by dfrankow**
This is a fixed-up copy of https://github.com/django/django/pull/1176/commits/4471a9f966ebbd564c4ad026c2bc0d99a95a2aa7, part of https://github.com/django/django/pull/1176, that claims to fix https://code.djangoproject.com/ticket/5711.

I don't know if we want this functionality anymore. The original ticket was filed 15 years ago.

However, if we do, here's something that claims to do it.

### CODE REVIEW

File: docs/topics/serialization.txt [hunk 1]
```diff
@@ -69,6 +69,29 @@ resulting output; it never appears in the ``fields`` part.
     serialized object doesn't specify all the fields that are required by a
     model, the deserializer will not be able to save deserialized instances.
 
+Extra attributes
+~~~~~~~~~~~~~~~~
+
+.. versionadded:: 4.2
```
Comment (dfrankow) [inline docs/topics/serialization.txt:75]: Not sure if this version is correct.

**Review by David-Wobrock — COMMENTED** [review]
Thanks for the PR!

However, I'm unsure Django is still maintaining the `.extra` method.
A few years back, in [#25213](https://code.djangoproject.com/ticket/25213) it was decided that the use of `.extra()` is discouraged, and that it won't receive anymore fixes.
From https://docs.djangoproject.com/en/dev/ref/models/querysets/#extra:
> This is an old API that we aim to deprecate at some point in the future. [...] We are no longer improving or fixing bugs for this method.

But maybe this patch has actually a wider scope than just the `.extra()` method 🤔 

We'll let the maintainers tell us on what should happen here, but maybe it's a _Won't Fix_ in the end :)

**Comment by dfrankow** [comment]
Thanks for looking.

Accepting this PR or closing the issue would be equal success to me. I just wanted to make something a bit better, and closing the issue would make things a bit better.

**Comment by dfrankow** [comment]
Also, any advice on how to get maintainers to look? I tried joining IRC, and never got a peep.

**Comment by dfrankow** [comment]
> Thanks for the PR!
> 
> However, I'm unsure Django is still maintaining the `.extra` method. A few years back, in [#25213](https://code.djangoproject.com/ticket/25213) it was decided that the use of `.extra()` is discouraged

This extra (from ticket 5711) is a different extra. This one is about serializing extra properties of an object. From the unit test:

>        serial_str = serializers.serialize(self.serializer_name, [a], extra=["age"])

The extra you mentioned is about passing different SQL to the ORM.

File: django/core/serializers/base.py [hunk 2]
```diff
@@ -125,8 +125,12 @@ def serialize(
         self.selected_fields = fields
         self.use_natural_foreign_keys = use_natural_foreign_keys
         self.use_natural_primary_keys = use_natural_primary_keys
+        self.extra_fields = options.pop("extra", None)
         progress_bar = self.progress_class(progress_output, object_count)
 
+        def field_is_selected(field):
```
Comment (David-Wobrock) [inline django/core/serializers/base.py:131]: Nitpicky: but I find it more readable when the name starts with the verb when the function returns a boolean.
```suggestion
        def is_field_selected(field):
```

File: django/core/serializers/base.py [hunk 3]
```diff
@@ -207,6 +205,14 @@ def handle_field(self, obj, field):
             "subclasses of Serializer must provide a handle_field() method"
         )
 
+    def handle_attr(self, obj, attr):
```
Comment (David-Wobrock) [inline django/core/serializers/base.py:208]: Did you mean `handle_extra_attr`?

File: docs/topics/serialization.txt [hunk 4]
```diff
@@ -69,6 +69,29 @@ resulting output; it never appears in the ``fields`` part.
...
+in addition to all the model fields.
+
+The ``fields`` and ``extra`` arguments can be combined to get a subset of the
+model fields as well as the extra non-field attributes specified.
+
+.. note::
+
+    Deserializing a model that was serialized with additional extra attributes
+    will result in an error, as deserialization requires that only the model
+    fields be specified.
```
Comment (David-Wobrock) [inline docs/topics/serialization.txt:93]: Hmm, that's a pity.

But would it work if we specified the `ignorenonexistent` param during deserialization? Like this `serializers.deserialize("xml", data, ignorenonexistent=True)`

File: tests/serializers/tests.py [hunk 5]
```diff
@@ -410,6 +410,17 @@ def test_serialize_inherited_fields(self):
...
+    def test_serialize_extra_non_field_attributes(self):
+        """
+        Tests that non-field attributes can be serialized if passed in
+        explicitly to fields.
+        """
+        a = Author(name="Some Random Person")
+        serial_str = serializers.serialize(self.serializer_name, [a], extra=["age"])
+        age_values = self._get_field_values(serial_str, "age")
+        self.assertEqual(len(age_values), 1)
+        self.assertEqual(int(age_values[0]), 42)
```
Comment (David-Wobrock) [inline tests/serializers/tests.py:422]: Maybe we can add some more tests? 🤔 
Specifying multiple extra fields, twice the same extra field, an extra field that doesn't exist - and then deserializing the data (with and without `ignorenonexistent`)

What do you think? 🙂

**Review by David-Wobrock — COMMENTED** [review]
Hey @dfrankow 

Yes, you're right, my bad! I went a bit fast over the PR, apologies 🙏 

The patch makes sense to me, I left some comments.
I assigned you on the [ticket](https://code.djangoproject.com/ticket/5711).
Once the comments are addressed, you can unset "Needs tests" and "Needs improvement" and the ticket will appear in the ones that require some review.

**Comment by dfrankow** [comment]
I would prefer to close this over submitting code, unless someone who is still around is actually requesting this feature.

The issue said "patch no longer merges cleanly", so I made it merge cleanly. However, I don't want this, and I don't know if anyone does. I also don't feel strongly about all the comments you just made, because I just ported over someone else's code without changing it. If someone wants this, then we can talk about the code quality.

**Comment by carltongibson** [comment]
Thanks for taking a look at this @dfrankow. Let's close given that you're not working on it. (It's here to pick up if someone wants to follow-up).

**Comment by dfrankow** [comment]
Thanks.

**Pull request closed without merging**